blog post about optimizing ingestion flow in pyroscope (#26)
pyroscope-io/docs: blog/profile-continuous-profiler.mdx

![push_vs_pull_diagram_07](https://user-images.githubusercontent.com/23323466/149289440-564c6e34-03cb-496d-bddc-faa96d075ced.gif)

One of the major benefits of pull mode is creating meaningful tags. For example, in Kubernetes workflows you can tag profiles with popular metadata fields:
- **pod_name**: Name of the pod
- **container_name**: Name of the container
- **namespace**: Namespace of the pod
- **service_name**: Name of the service
- etc

Then using Pyroscope's query language, [FlameQL](/docs/flameql), you can filter profiles by those tags over time to see where the application is spending most of its time.

Early on we had a user who ran pyroscope in pull mode with about a thousand of profiling targets and about **1 TB of raw profiling data per day**, but their Pyroscope server was running into some performance issues. 

This was surprising to us because in push mode, we've seen pyroscope handle similar amounts of traffic without issues. Therefore, we suspected that the performance issue had to do with one of the key architectural difference between push mode and pull mode. 

> While `pprof` is a great format for representing profiling data, but it's not the most optimal format when it comes to storing profiling data. So, we transcode `pprof` into our internal format that we optimized for storage in our [custom storage engine](/docs/storage-design).

Author: ```suggestion
While `pprof` is a great format for representing profiling data, it's not the most optimal format when it comes to storing profiling data. So, we transcode `pprof` into our internal format that we optimized for storage in our [custom storage engine](/docs/storage-design).
```

- **Push mode**: In push mode pprof transcoding is done on each profiling target so the load was distributed across many targets
- **Pull mode**: In pull mode we moved pprof transcoding to the server side so the load was distributed across only one target

So, we immediately suspected that this performance regression might have something to do with the load increase on the pulling server.

> ![pprof_transcoding_04](https://user-images.githubusercontent.com/23323466/150457067-3f62aa87-e05b-48d2-821e-6d2770155f9c.gif)

Author: ```suggestion

![pprof_transcoding_04](https://user-images.githubusercontent.com/23323466/150457067-3f62aa87-e05b-48d2-821e-6d2770155f9c.gif)
```

add a new line so that there's padding between text and image

<div style={{ width: "100%", margin: "auto" }}>

![third_fg_image](https://user-images.githubusercontent.com/23323466/150475638-d85225d1-030c-4112-ba11-50a0b9e570b1.png)
</div>

<center style={{margin:"-15px 20px 20px"}}><i>The green nodes are functions where we decreased CPU usage and the red nodes are functions where we added CPU usage for this optimization. We removed <code>FindName</code> and <code>FindLocation</code> and also removed <code>runtime.mapaccess2_fast64</code></i></center>

## What happens when you profile a profiler?

In retrospect, we probably should have _started_ with looking at the specifications for a pprof object before assuming that binary search was the best way to find objects.
But we wouldn't have even know to look at the specifications until we started profiling our own code.

It's just the nature of developing software that as complexity increases over time, more performance issues sneak their way into the codebase. 

However, with continuous profiling enabled, it allows you to spot these performance issues in your code and understand which parts of your code are consuming the most resources. 

> If you'd like to learn more about how to get started with Pyroscope and learn where your code's bottleneck is, check out the [Pyroscope documentation](/docs/installing-pyroscope-overview).

Author: ```suggestion
If you'd like to learn more about how to get started with Pyroscope and learn where your code's bottlenecks are, check out the [Pyroscope documentation](/docs/installing-pyroscope-overview).
```

Author: i don't feel particularly strongly about this one
